File: julia/day_02.jl
"Load data from file."
function loaddata(file::String)
    return readlines(file)
end

"Check password according to sled rental policy."
function checksled(low::Int, up::Int, char::Char, password::AbstractString)
    return count(string(char), password) ∈ low:up
end

"Check password according to toboggan rental policy."
function checktoboggan(low::Int, up::Int, char::Char, password::AbstractString)
    return (password[low] == char) ⊻ (password[up] == char)
end

"Count valid passwords according to given policy."
function countvalid(data::Array, rule::Function)
    n_count = 0
    for line in data
        low, up, char, password = split(replace(line, r"[-:]" => " "))
        n_count += rule(parse(Int, low), parse(Int, up), char[1], password)
    end
    return n_count
end

function main()
    data = loaddata("../assets/data/02.txt")

    # part one
    n_valid = countvalid(data, checksled)
    println("The product of the two numbers is ", n_valid, ".")

    # part two
    n_valid = countvalid(data, checktoboggan)
    println("The product of the two numbers is ", n_valid, ".")
end

main()

File: 2020/python/day_02.py
"""Day 2: Password Philosophy"""


def data_from(file_: "str") -> list:
    """Returns data from given file."""
    data = []
    with open(file_, "r") as f:
        for line in f.readlines():
            data.append(line.replace("-", " ").replace(":", "").split())
    return data


def count(char: "str", pw: "str") -> int:
    """Count character in password."""
    n_match = 0
    for x in pw:
        if char == x:
            n_match += 1
    return n_match


def check_valid(policy_and_pw: "list") -> bool:
    """Check whether password policy is met or not."""
    min_, max_, char, pw = policy_and_pw
    n_match = count(char, pw)
    if n_match < int(min_) or n_match > int(max_):
        return False
    else:
        return True


def check_valid_new(policy_and_pw: "list") -> bool:
    """Check whether new password policy is met or not."""
    pos0, pos1, char, pw = policy_and_pw
    if (pw[int(pos0) - 1] == char) + (pw[int(pos1) - 1] == char) == 1:
        return True
    else:
        return False


def main():
    file_ = "../assets/data/02.txt"
    data = data_from(file_)

    # part one
    n_valid = 0
    for policy_and_pw in data:
        if check_valid(policy_and_pw):
            n_valid += 1

    print(f"A total of {n_valid} passwords are valid according to their policies.")

    # part two
    n_valid = 0
    for policy_and_pw in data:
        if check_valid_new(policy_and_pw):
            n_valid += 1

    print(f"A total of {n_valid} passwords are valid according to the new policies.")


if __name__ == "__main__":
    main()
